In [58]:
import pandas as pd
import numpy as np

# Data Collection

In [59]:
data = pd.read_excel("/content/Online Retail.xlsx", sheet_name ='Online Retail')

In [60]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

# Exploring Data

In [61]:
data.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Columns: 8 entries, InvoiceNo to Country
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [62]:
#Percentage of null values in all columns
data.isnull().sum()/len(data)*100

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64

In [63]:
data.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

In [64]:
data.value_counts("StockCode")

StockCode
85123A    2313
22423     2203
85099B    2159
47566     1727
20725     1639
          ... 
84031b       1
84031a       1
82613a       1
20950        1
m            1
Length: 4070, dtype: int64

# Data Preprocessing

*Handling Null Values*






In [65]:
data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Exploring CustomerID Feature

In [66]:
data.value_counts("CustomerID")

CustomerID
17841.0    7983
14911.0    5903
14096.0    5128
12748.0    4642
14606.0    2782
           ... 
14119.0       1
13747.0       1
15590.0       1
15562.0       1
17925.0       1
Length: 4372, dtype: int64

In [67]:
data[data['CustomerID'].isna()==True]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
622    2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443   2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444   2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445   2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446   2010-12-01 14:32:00       1.66         NaN  United Kingdom  
...                    ...        ...         ...             ...  
541536 2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541537 2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541538 2011-12-09 10:26:00       4.96         NaN  United Kingdom  
541539 2011-12-09 10:26:00      10.79         NaN  United Kingdom  
541540 2011-12-09 10:26:00    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

In [68]:
distinct_counts = data.groupby('InvoiceNo')['CustomerID'].nunique().reset_index()

In [69]:
distinct_counts[distinct_counts['CustomerID']==0]

InvoiceNo  CustomerID
43       536414           0
84       536544           0
85       536545           0
86       536546           0
87       536547           0
...         ...         ...
25661   C578097           0
25729   C579757           0
25769   C579907           0
25816   C580604           0
25817   C580605           0

[3710 rows x 2 columns]

In [70]:
distinct_counts[distinct_counts['CustomerID']>1]

Empty DataFrame
Columns: [InvoiceNo, CustomerID]
Index: []

In [71]:
data.groupby('InvoiceDate').agg({'InvoiceNo': 'unique', 'CustomerID': 'unique'}).reset_index()

InvoiceDate         InvoiceNo CustomerID
0     2010-12-01 08:26:00          [536365]  [17850.0]
1     2010-12-01 08:28:00          [536366]  [17850.0]
2     2010-12-01 08:34:00  [536367, 536368]  [13047.0]
3     2010-12-01 08:35:00          [536369]  [13047.0]
4     2010-12-01 08:45:00          [536370]  [12583.0]
...                   ...               ...        ...
23255 2011-12-09 12:23:00          [581583]  [13777.0]
23256 2011-12-09 12:25:00          [581584]  [13777.0]
23257 2011-12-09 12:31:00          [581585]  [15804.0]
23258 2011-12-09 12:49:00          [581586]  [13113.0]
23259 2011-12-09 12:50:00          [581587]  [12680.0]

[23260 rows x 3 columns]

**Updating CustomerID Null values**

There are total 3710 unique invoices for which there is no CustomerID hence we cannot drop these rows as it will result in loss of lots of relevant data Hence I will replace it with some Customer ID based on Invoice Number.

*The replaced CustomerID will start with "RE"*

In [72]:
#Missing values in the "CustomerID" column replaced based on the values within the same "InvoiceNo" group
data['CustomerID'] = data.groupby('InvoiceNo')['CustomerID'].transform(lambda x: x.fillna(method='ffill').fillna(method='bfill'))

In [74]:
data[data['CustomerID'].isna()==True]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
622    2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443   2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444   2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445   2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446   2010-12-01 14:32:00       1.66         NaN  United Kingdom  
...                    ...        ...         ...             ...  
541536 2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541537 2011-12-09 10:26:00       4.13         NaN  United Kingdom  
541538 2011-12-09 10:26:00       4.96         NaN  United Kingdom  
541539 2011-12-09 10:26:00      10.79         NaN  United Kingdom  
541540 2011-12-09 10:26:00    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

As we have explored above that there is no invoiceNo for which there are 2 distinct values of CustomerId Hence the count is still the same.

*So, we need to update the customerId with new values*

In [73]:
invoices_with_empty_customer_id = data[data['CustomerID'].isna()]['InvoiceNo'].unique()

In [75]:
#counting distinct invoice numbers having null customerID for replacing Null CustomerId with value
len(invoices_with_empty_customer_id)

3710

In [76]:
new_customer_id = "RE0001"
customer_count = 1
invoice_customer_mapping = {}
for invoice_no in invoices_with_empty_customer_id:
  if(invoice_no not in invoice_customer_mapping):
    data.loc[data['InvoiceNo'] == invoice_no, 'CustomerID'] = new_customer_id
    invoice_customer_mapping[invoice_no] = new_customer_id
    customer_count += 1
    new_customer_id = "RE" + str(customer_count).rjust(4,"0")
  else:
    data.loc[data['InvoiceNo'] == invoice_no, 'CustomerID'] = invoice_customer_mapping[invoice_no]

In [77]:
data[data['CustomerID'].isna()==True]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

Handling Description Feature

In [78]:
data[data['Description'].isna()==True]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
622       536414     22139         NaN        56 2010-12-01 11:52:00   
1970      536545     21134         NaN         1 2010-12-01 14:32:00   
1971      536546     22145         NaN         1 2010-12-01 14:33:00   
1972      536547     37509         NaN         1 2010-12-01 14:33:00   
1987      536549    85226A         NaN         1 2010-12-01 14:34:00   
...          ...       ...         ...       ...                 ...   
535322    581199     84581         NaN        -2 2011-12-07 18:26:00   
535326    581203     23406         NaN        15 2011-12-07 18:31:00   
535332    581209     21620         NaN         6 2011-12-07 18:35:00   
536981    581234     72817         NaN        27 2011-12-08 10:33:00   
538554    581408     85175         NaN        20 2011-12-08 14:06:00   

        UnitPrice CustomerID         Country  
622           0.0     RE0001  United Kingdom  
1970          0.0     RE0003  United Kingdom  
1971          0.0     RE0004  United Kingdom  
1972          0.0     RE0005  United Kingdom  
1987          0.0     RE0006  United Kingdom  
...           ...        ...             ...  
535322        0.0     RE3681  United Kingdom  
535326        0.0     RE3685  United Kingdom  
535332        0.0     RE3691  United Kingdom  
536981        0.0     RE3699  United Kingdom  
538554        0.0     RE3703  United Kingdom  

[1454 rows x 8 columns]

In [85]:
stocks_with_empty_description = data[data['Description'].isna()]['StockCode'].unique()
len(stocks_with_empty_description)

960

In [80]:
d_count = data.groupby('StockCode')['Description'].nunique().reset_index()

In [83]:
d_count[d_count['Description']==0]

StockCode  Description
5         10134            0
26        16053            0
80        20689            0
111       20738            0
159       20825            0
...         ...          ...
4046   DCGS0057            0
4047  DCGS0066P            0
4052   DCGS0071            0
4053   DCGS0072            0
4055   DCGS0074            0

[112 rows x 2 columns]

In [86]:
#Missing values in the "Description" column replaced based on the values within the same "StockCode" group
data['Description'] = data.groupby('StockCode')['Description'].transform(lambda x: x.fillna(method='ffill').fillna(method='bfill'))

In [87]:
data[data['Description'].isna()==True]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
1970      536545     21134         NaN         1 2010-12-01 14:32:00   
1987      536549    85226A         NaN         1 2010-12-01 14:34:00   
1988      536550     85044         NaN         1 2010-12-01 14:34:00   
2024      536552     20950         NaN         1 2010-12-01 14:34:00   
2026      536554     84670         NaN        23 2010-12-01 14:35:00   
...          ...       ...         ...       ...                 ...   
280754    561498     21610         NaN       -14 2011-07-27 14:10:00   
281615    561555    37477B         NaN       -11 2011-07-28 10:21:00   
281616    561557    37477C         NaN       -31 2011-07-28 10:21:00   
346849    567207    35592T         NaN         4 2011-09-19 11:01:00   
497301    578360    84971L         NaN         2 2011-11-24 10:36:00   

        UnitPrice CustomerID         Country  
1970          0.0     RE0003  United Kingdom  
1987          0.0     RE0006  United Kingdom  
1988          0.0     RE0007  United Kingdom  
2024          0.0     RE0008  United Kingdom  
2026          0.0     RE0010  United Kingdom  
...           ...        ...             ...  
280754        0.0     RE2380  United Kingdom  
281615        0.0     RE2404  United Kingdom  
281616        0.0     RE2405  United Kingdom  
346849        0.0     RE2754  United Kingdom  
497301        0.0     RE3530  United Kingdom  

[112 rows x 8 columns]

Still there are 112 rows having no description value

We can notice that most of these rows have unit price 0, hence they are contributing less to revenue so we can drop these rows

In [91]:
data = data.dropna(subset=["Description"])

In [92]:
data[data['Description'].isna()==True]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

*All the columns are in proper format hence no data type conversion is necessary*

# Statistical Analysis